File: MetOpt/metopt_lab1/src/main/java/com/example/metopt/math/expression/operations/Variable.java
package com.example.metopt.math.expression.operations;

import com.example.metopt.math.expression.Expression;
import com.example.metopt.math.expression.exception.UndefinedVariableException;

import java.util.EnumMap;

/**
 * @author Danil Demintsev (demintsievd@yandex.ru) on 26.02.2021
 */

public class Variable implements Expression {

    private final VariableName name;

    public Variable(VariableName name) {
        this.name = name;
    }

    @Override
    public double evaluate(final EnumMap<VariableName, Double> vars) {
        Double value = vars.get(name);
        if (value == null) {
            throw new UndefinedVariableException("Variable " + name.name() + " is not defined");
        }
        return value;
    }

    @Override
    public String toString() {
        return name.token;
    }
}

File: MetOpt/metopt_lab2-master/expression/src/expression/nullary/Variable.kt
package expression.nullary

import expression.Expression

/**
 * @author Danil Demintsev (demintsievd@yandex.ru) on 26.02.2021
 */
enum class Variable(
    val token: String,
) : Expression {

    X("x"),
    Y("y"),
    Z("z");
    //TODO: x1, x2, x3, ...

    override fun evaluate(vars: Map<Variable, Double>): Double =
        vars[this] ?: throw IllegalArgumentException("No value for variable with name $this")

    override fun diff(variable: Variable): Expression =
        if (variable == this) {
            Const.ONE
        } else {
            Const.ZERO
        }

    override fun toString(): String = token
}
